### PR TITLE
Add empty jid reply processing

### PR DESCRIPTION
**Pull request by ezh**
Related to #141

### CODE REVIEW

File: pepper/cli.py [hunk 1]
```diff
@@ -540,6 +540,22 @@ def run(self):
...
+            while True:
+                total_time = time.time() - start_time
+                if total_time > self.options.timeout:
+                    exit_code = 1
+                    break
+                try:
+                    ret = self.low(api, load)
+                    if all(ret["return"][0].values()):
+                        break
+                    logger.debug('Incorrect reply {}'.format(str(ret)))
```
Comment (gtmanfred) [inline pepper/cli.py]: please use the string logger format, it is faster.

```
logger.debug('Incorrect reply: %s', ret)
```

File: pepper/cli.py [hunk 2]
```diff
@@ -540,6 +540,22 @@ def run(self):
...
+                except:
```
Comment (gtmanfred) [inline pepper/cli.py]: Please specify `except Exception`

File: pepper/cli.py [hunk 3]
```diff
@@ -540,6 +540,22 @@ def run(self):
...
+                    logger.debug('Error with self.low(api, load) request', exc_info=True)
+                    pass
```
Comment (gtmanfred) [inline pepper/cli.py]: This pass is not needed

**Review by gtmanfred — CHANGES_REQUESTED** [review]

**Review by gtmanfred — APPROVED** [review]

**Comment by gtmanfred** [comment]
Can you give me more information on what the goal here is?

If you want it to use poll_for_return, then you should be passing in `--fail-if-incomplete`

**Comment by gtmanfred** [comment]
What is happening here looks like it ends up calling the same api call over and over again if it gets that jid return, which is not desirable.

**Comment by ezh** [comment]
This PR use pepper option
-t TIMEOUT, --timeout=TIMEOUT
                         Specify wait time (in seconds) before returning
                        control to the shell

File: pepper/cli.py [hunk 4]
```diff
@@ -607,6 +607,21 @@ def run(self):
...
+            # keep trying until all expected nodes return
+            total_time = 0
+            start_time = time.time()
+            while True:
+                total_time = time.time() - start_time
+                if total_time > self.options.timeout:
+                    exit_code = 1
+                    break
+                try:
+                    ret = self.low(api, load)
```
Comment (gtmanfred) [inline pepper/cli.py:620]: but it is making the same api call over and over again, which I don't think is desirable, you could potentially try to run a highstate twice in a row.

Comment (ezh) [reply pepper/cli.py:620]: Yep. Nice catch.

**Comment by gtmanfred** [comment]
Overall I think this is a good idea, but I would really like to have it be opt-in, instead of the default.

Something they have to say, `I want to run async, and automatically query for the results in the job database`

File: pepper/cli.py (hunk 4 above)
Comment (ezh) [reply pepper/cli.py:620]: No. SaltStack strictly prohibit parallel calls of the same states by default. https://docs.saltstack.com/en/latest/ref/states/parallel.html It is possible if you explicitly allow parallel processing.
https://stackoverflow.com/questions/42972212/saltstack-run-parallel-wait-for-other-jobs or pass queue argument.

Comment (ezh) [reply pepper/cli.py:620]: But I understood your concern. It depends on environment.

Comment (gtmanfred) [reply pepper/cli.py:620]: Yes, but that doesn't mean that any other command could be run more than once.

What if you are trying to run a http.query, and the minion just starts running 10 commands that download a 1G file.

**Comment by ezh** [comment]
Yep. It is up to you. It looks that I just changed behaviour, not fix the issue.

**Comment by gtmanfred** [comment]
I am going to leave this PR as closed.

But am still interested in trying to fix the issue that you opened.

**Pull request closed without merging**